### PR TITLE
stm32 i2c: fix master mode reads, and add support for the NXP SE050 crypto device.

### PR DESCRIPTION
**Pull request by ldts**

### CODE REVIEW

**Review by jforissier — COMMENTED** [review]
Edit: For "drivers: stm32_i2c: fix read operations on I2C_MODE_MASTER mode" only:

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by ldts** [comment]
tag added to "drivers: stm32_i2c: fix read operations on I2C_MODE_MASTER mode"

**Comment by ldts** [comment]
@etienne-lms maybe you can check my second patch please?

**Comment by ldts** [comment]
would be nice to have this merged so that STM32MP1 is supported with the SE050 device; the platform could then use the cryptographic driver with the pkcs11 support by just plugging one  of [these](https://www.nxp.com/products/security-and-authentication/authentication/edgelock-se051-development-kit:OM-SE051ARD) to the I2C bus

File: core/drivers/crypto/se050/glue/i2c_stm32.c [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+	return TEE_SUCCESS;
+}
+
+static int dt_i2c_bus_config(struct stm32_i2c_init_s *init,
+			     struct stm32_pinctrl **pctrl,  size_t *pcnt)
+{
+	const fdt32_t *cuint = NULL;
+	const char *path = NULL;
+	char bus[6] = { '\0' };
```
Comment (jforissier) [inline core/drivers/crypto/se050/glue/i2c_stm32.c:38]: `= { };` is preferred

File: core/drivers/crypto/se050/glue/i2c_stm32.c [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+ * Copyright (C) Foundries Ltd. 2022 - All Rights Reserved
+ * Author: Jorge Ramirez-Ortiz <jorge@foundries.io>
+ */
+
+#include <drivers/stm32_i2c.h>
+#include <i2c_native.h>
+#include <kernel/boot.h>
+#include <kernel/dt.h>
+#include <libfdt.h>
+#include <phNxpEsePal_i2c.h>
```
Comment (jforissier) [inline core/drivers/crypto/se050/glue/i2c_stm32.c:12]: Where is this file supposed to be?

File: core/drivers/crypto/se050/glue/i2c_stm32.c [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+{
+	struct stm32_i2c_init_s i2c_init = { };
+	struct stm32_pinctrl *pinctrl = NULL;
+	size_t pin_count = 0;
+
+	/* No need to re-init */
+	if (i2c.base.pa)
+		return 0;
+
+	/* support only one device on the platform */
```
Comment (jforissier) [inline core/drivers/crypto/se050/glue/i2c_stm32.c:75]: Uppercase S preferred

File: core/drivers/crypto/se050/glue/i2c_stm32.c [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+	if (dt_i2c_bus_config(&i2c_init, &pinctrl, &pin_count))
+		return -1;
+
+	/* probe the device */
```
Comment (jforissier) [inline core/drivers/crypto/se050/glue/i2c_stm32.c:79]: Uppercase P

File: core/drivers/stm32_i2c.c [hunk 5]
```diff
@@ -1383,6 +1383,10 @@ static int i2c_read(struct i2c_handle_s *hi2c, struct i2c_request *request,
 	if (i2c_wait_stop(hi2c, timeout_ref))
 		goto bail;
 
+	/* Clear the NACK generated at the end of the transfer */
+	if ((io_read32(get_base(hi2c) + I2C_ISR) & I2C_ISR_NACKF))
+		io_write32(get_base(hi2c) + I2C_ICR, I2C_ICR_NACKCF);
```
Comment (etienne-lms) [inline core/drivers/stm32_i2c.c:1388]: This looks good to me but I've asked a 2nd eye on this. Please stay tune, I'll answer asap.

File: core/drivers/crypto/se050/glue/i2c_stm32.c (hunk 2 above)
Comment (ldts) [reply core/drivers/crypto/se050/glue/i2c_stm32.c:12]: it is in the Plug-and-Trust repo (it is an external dependency)
https://github.com/foundriesio/plug-and-trust

File: core/drivers/stm32_i2c.c (hunk 5 above)
Comment (ldts) [reply core/drivers/stm32_i2c.c:1388]: thanks @etienne-lms

**Comment by ldts** [comment]
@etienne-lms added an extra commit to improve the overall read performance.

File: core/drivers/stm32_i2c.c (hunk 5 above)
Comment (ldts) [reply core/drivers/stm32_i2c.c:1390]: perhaps we should just clear it unconditionally

Comment (ldts) [reply core/drivers/stm32_i2c.c:1390]: @etienne-lms just a reminder that this fix is pending - ie, the current driver upstream is broken in this xfer mode.

**Comment by ldts** [comment]
@jforissier if we could pick these fixes and features before next release that would be awesome.

**Comment by jforissier** [comment]
Sure.

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

I will wait a bit before merging in case @etienne-lms has more comments but it shouldn't be a problem to have that in the next release.

**Comment by etienne-lms** [comment]
Hi, sorry for the late feedback.

For commit 'drivers: stm32_i2c: fix read operations on I2C_MODE_MASTER mode':
I'm still discussing with our expert and we want to understand better the situation. I've agreed to get this change merged before next OP-TEE tag if we don't get to a conclusion in the mean time. Please, give us few more days.

`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` for commit 'se050: glue: i2c_stm32'.
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for commit 'drivers: stm32_i2c: improve read transfer speed' but that commit header line should be rephrased (IMO): it speeds up error detection, not read transfers themselves.

**Comment by jforissier** [comment]
@etienne-lms understood, thanks

**Review by etienne-lms — COMMENTED** [review]
Oops, I forgot to post these 2 minor comments.

**Comment by jforissier** [comment]
@ldts would you mind adding my Acked-by to the last two commits? Then I will merge this. Thanks!

**Comment by etienne-lms** [comment]
Fine with me. Sorry I have no feedback so far, ok for merging. Thanks @jforissier.

**Pull request merged**